File: src/ocaml/domain/node.ml
(** Node is a domain in a file tree. *)
open Sxfiler_core

(** identifier in file tree. *)
type id = string [@@deriving eq, show]

(** [link_path] will have target of the link if item pointed to full_path is *link. *)
type t =
  { id : id
  ; full_path : Path.t
  ; stat : File_stat.t
  ; link_path : Path.t option }
[@@deriving eq, show, fields]

let has_same_id v1 v2 = equal_id v1.id v2.id
let make ~id ~full_path ~stat ~link_path = {id; full_path; stat; link_path}

(** [is_directory v] shortcut function to detect node is directory or not *)
let is_directory v = v.stat.File_stat.is_directory

File: src/ts/domain/node.ts
// define type and operations for node
import { FileStat } from "./file-stat";

export class Node {
  /**
   * create new Node
   * @param id id of node
   * @param name file or directory name of node
   * @param stat stats of node
   * @param parentDirectory parent directory of node
   * @param linkPath link target path if node is link
   */
  constructor(
    public readonly id: string,
    public readonly name: string,
    public readonly stat: FileStat,
    public readonly parentDirectory: string,
    public readonly linkPath?: string
  ) {}
}
